### ISSUE TITLE
Nullish coalescing short-circuits type checking (only in strict mode)

### ISSUE THREAD

**Issue by travigd** [microsoft/TypeScript#36393]
```
$ ./node_modules/.bin/tsc --version
Version 3.7.2
```

**Search Terms:**

**Code**

A toy example would be something like this.
```ts
let foo: string = "";
foo = "bar" ?? 123;
```
This one is obviously fine since `"bar"` is always truthy.

However, this becomes a little bit problematic when you consider the idiom of having `Record` objects and checking their truthiness before using them.
```
const elts = ["foo", "bar", "spam", "spam", "foo", "eggs"];
const counts: Record<string, number>;
for (const elt of elts) {
  // This really **should** raise an error.
  counts[elt] = counts[elt] ?? "zero";
  counts[elt] += 1;
}
```

**Expected behavior:**
An error should be raised.

**Actual behavior:**
Curiously, an error is **not** raised in strict mode but **is** raised in un-strict mode.

```sh
$ ./node_modules/.bin/tsc ./foo.ts 
foo.ts:5:3 - error TS2322: Type 'number | "zero"' is not assignable to type 'number'.
  Type '"zero"' is not assignable to type 'number'.

5   counts[elt] = counts[elt] ?? "zero";
    ~~~~~~~~~~~


Found 1 error.
```

```sh
$ ./node_modules/.bin/tsc --strict ./foo.ts
# No error, exits 0 and emits JS.
```
**Playground Link:** 
[Playground Link](http://www.typescriptlang.org/play/?strictNullChecks=false&ts=3.8.0-dev.20200122#)

Toggling the `strictNullChecks` config option will show the issue.

**Related Issues:**

**Comment by jcalz** [comment]
This is probably related to or a duplicate of #13778.  

With `--strictNullChecks` on, the compiler assumes that `counts[elt]` must be of type `number`, which is never `null` or `undefined`.  If you want the compiler to notice that `counts[elt]` may be `undefined`, then you (currently) need to include `undefined` in the property type manually:

```ts
const counts: Record<string, number | undefined> = {};
for (const elt of elts) {
  counts[elt] = (counts[elt] ?? "zero"); // error now
  counts[elt] = (counts[elt] ?? 0) + 1; // okay
}
```
[Playground](http://www.typescriptlang.org/play/?ts=3.8.0-dev.20200122#code/MYewdgzgLgBApgGyhGBeGBtARAMxCLAGhiwCMBDAJyJIgAdyBbGreplvA4rOAc14hYAugG4AUKEixQAVzDIAXDABKcUJQAmAHmiUAlmF7EwMxqTiUYAHxhyNcHAbgaAfGhgBvAL7i8lgBSS0PBIMCA4IcgAlJ5iMDCy8hAYiFBC7oEgcsgpSOkA-PkkAF4WBFEiMAD0VfCUlCCWYCAA7nEJWUm5aRmJOakFRQAMMQDUMACMlTVhANbkAJ5iXkA)

With `--strictNullChecks` off, the compiler never assumes nullish coalescing will short-circuit because it can never eliminate `null` and `undefined`.  See [this comment from the PR](https://github.com/microsoft/TypeScript/pull/32883#discussion_r314608914)

**Comment by travigd** [comment]
It seems like there could be some "rightward-precedence" for this kind of thing (I'm not sure about the right terminology to describe what I mean).

What I mean is basically this:
```
string -> string
string ?? number -> string | number;
null ?? number -> number;
```

I think the algorithm would roughly be something like
* Given a chain of `A ?? B ?? C ?? ...`, take the first pair (nullish coalescing is left-associative, right?)
* Resolve the types of `A` and `B`
* Judge the type of `A ?? B` to be `NonNull<A> | B`
* Recurse

```ts
string ?? (string | null) ?? string[]
// Take first pair...
=> (NonNull<string> | (string | null)) ?? string[]
// Simplify the types...
=> (string | null) ?? string[]
// Recursively take first pair...
=> NonNull<string | null> | string[]
// Simplify the types...
=> string | string[]
// Done!
```

**Comment by falsandtru** [comment]
`&&`, `||`, and `!` operators also have a similar problem.

**Comment by kherock** [comment]
I just ran into this myself, but I'd say this is working as intended. Rephrasing what @jcalz found, the example in the issue description is demonstrating an example of a `Record` type that is too loose:

```typescript
const counts: Record<string, number>;

// strictNullChecks: true
counts.foo // evaluated as (number)

// strictNullChecks: false
counts.foo // evaluated as (number | null | undefined)
```

Under strict mode, [@typescript-eslint/no-unnecessary-condition](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-unnecessary-condition.md) does catch TS's short-circuiting behavior when computing the type and will yell at you:

```typescript
const elts = ['foo', 'bar', 'spam', 'spam', 'foo', 'eggs'];
const counts: Record<string, number> = {};
for (const elt of elts) {
  counts[elt] = counts[elt] ?? 'zero'; /*
                ^^^^^^^^^^^
Unnecessary conditional, expected left-hand side of `??` operator to be possibly null or undefined. [eslint] */
  counts[elt] += 1;
}
```

If you explicitly widen the type to include `null | undefined`, you'll get the error you see in loose mode:

```typescript
const elts = ['foo', 'bar', 'spam', 'spam', 'foo', 'eggs'];
const counts: Record<string, number> = {};
for (const elt of elts) {
  counts[elt] = (counts[elt] as number | null | undefined) ?? 'zero'; /*
  ^^^^^^^^^^^
Type 'number | "zero"' is not assignable to type 'number'.  Type 'string' is not assignable to type 'number'. [ts] */
  counts[elt] += 1;
}
```

Given that a linter is capable of pointing out the potentially unexpected behavior, I don't know if there's anything that needs to be fixed on TS's end.

**Comment by travigd** [comment]
I respectfully disagree (obviously - I created the issue 😉). Using too-loose `Record` types is a common idiom in TypeScript, and I think changing this behavior makes that idiom easier to work with.

**Comment by dmo-odoo** [comment]
I think the following is related to this issue:

```ts
function auto<T1, T2>(f: () => T1, fallback: T2) {
    return f() ?? fallback;
}
// Incorrectly typed to "number | undefined" even though the right hand path of `??` will never be taken in this case.
const f1 = auto(() => 42, undefined);
function forced<T1, T2>(f: () => T1, fallback: T2): T1 extends undefined | null ? NonNullable<T1> | T2 : T1{
    return f() ?? fallback as any;
}
// Correctly typed to "number".
const f2 = forced(() => 42, undefined);
```

This can be important if your system relies on fallback values. With the current typing of `??`, the system would force you to define a T2 fallback because if you don't define one then its `undefined` typing from being optional will pollute the final return type even in cases where T1 can never be `null` or `undefined`.

**Comment by duddlf23** [comment]
It seems to be resolved by turning on noUncheckedIndexedAccess config.
